### PR TITLE
[DO NOT MERGE] RavenDB-5291, RavenDB-5006

### PR DESCRIPTION
**Pull request by redknightlois**

### CODE REVIEW

**Comment by redknightlois** [comment]
DO NOT MERGE!!! Rebase issues killed the PR.

File: src/Voron/Constants.cs [hunk 1]
```diff
@@ -46,10 +47,13 @@ public static class Size
         public const int CurrentVersion = 7;
 
         public const string RootTreeName = "$Root";
+        public static readonly Slice RootTreeNameSlice = Slice.From(StorageEnvironment.LabelsContext, RootTreeName, ByteStringType.Immutable);
```
Comment (ayende) [inline src/Voron/Constants.cs:50]: how thread safe is this? If we have two threads initializing two different static fields using the same label context?

**Comment by redknightlois** [comment]
Completely thread-safe. Statics are instantiated at class loading time in the loader context which is guaranteed to run in a single thread.

**Comment by ayende** [comment]
No, they are not.
See: https://msdn.microsoft.com/en-us/library/aa645758(v=vs.71).aspx

Imagine two separate classes using this.
Leaving aside that different fields on the same class can be init by different threads.

**Comment by redknightlois** [comment]
Not an small detail. You are right, we need to move the initialization into the static constructor. I plan to get  rid of the entire issue (I already discussed what we are going to do with Julian already) but in the meantime I prefer to move them to the static constructor.

cc @jbayardo

**Comment by ayende** [comment]
Closing because it is currently conflicting and we need a new one.

**Comment by redknightlois** [comment]
Julian has been sick since Saturday, so that's why this one is still conflicting.

**Pull request closed without merging**